### ISSUE TITLE
docker containers cannot use virtual domain

### ISSUE THREAD

**Issue by GuliyevFarid** [jwilder/nginx-proxy#1397]
I created a new docker container of nginx-proxy. I added virtual_host and virtual_port to all my containers. Currently all my containers are in the same docker network.

I have faced with the following issue. Host cannot reach containers by virtual_host (domain). So, I have added the following record to my hosts file and next I could reach docker containers via domains.

for example, my "hosts" file looks like that:

> 127.0.0.1 server1.local
> 127.0.0.1 server2.local

I am able to reach containers from the host. But docker container server1 cannot reach server2.local.
As I added server2.local to my hosts file, so, server1 also uses my hosts file by default and the request to server2.local is forwarded to localhost which is wrong.

so, the reverse proxy does not support containers to see each other by virtual_host.

**Comment by sgabe** [comment]
Note that containers on the default bridge network can only access each other by IP addresses. Are you using a user-defined bridge network?

**Comment by GuliyevFarid** [comment]
Yes, I use custom network with bridge driver.

**Comment by sgabe** [comment]
The title does not reflect the actual issue, which is much more a general question about how to configure or use Docker and not specific to nginx-proxy.

In a user-defined `bridge` network the container’s network stack is isolated from the Docker host. Based on the provided information, it seems that your containers are using `host` networking. In that case the `/etc/hosts` file will be the same on the host and in the containers, due to the fact that the containers share the host’s networking namespace.

**Comment by GuliyevFarid** [comment]
Hi @sgabe ,  

here is the list of networks
![image](https://user-images.githubusercontent.com/23084008/74608800-7b52a200-50e4-11ea-8381-6377351044d8.png)

and below screenshot shows the containers
![image](https://user-images.githubusercontent.com/23084008/74608754-2b73db00-50e4-11ea-9df8-67081a6cf399.png)

looking to ip addresses, it is clear that containers are in the user-defined bridge network.

I do not know if user-defined bridge network should be able to use hosts file from the host or not, but it is as it is, I mean containers can use/ping the domain that is written in hosts from the host.

**Comment by sgabe** [comment]
User-defined bridges provide automatic DNS resolution between containers, so containers in 172.18.0.0/16 should be able to resolve each other by name or alias, by default. You might be using the `--add-host` option with `docker run` to add lines to `/etc/hosts` in the containers. I am not sure how this issue is related to the nginx-proxy project.

**Comment by AndrewTriesToCode** [comment]
@sgabe I agree, but nginx-proxy configures the /etc/nginx/conf.d/default.conf file to route "upstream" via IP address so it doesn't explain why it doesn't work. For some reason it doesn't work on the default network. If you use your own bridge network it works fine--even though in both cases nginx is using the IP addresses internally.

**Comment by tkw1536** [comment]
@GuliyevFarid I encountered a similar issue in the past. In particular, I had a problem when the name of a container was equal to the `VIRTUAL_HOST`. I'm not sure if you have the same issue, but I'll write this comment just in case.  

Assuming the following:
- container `172.18.0.2` is named `server1.local`, `VIRTUAL_HOST=server1.local` and `VIRTUAL_PORT=8080`
- container `172.18.0.3` is on the same network as `172.18.0.2`

Now what happens if 172.18.0.3 attempts to connect to `http://server1.local`?
1. It will try to resolve `server1.local`. 
2. Because of automatic DNS resolution, it will find the IP address `172.18.0.2`
3. The container will attempt to connect to `172.18.0.3` on port 80 to retrieve the website. 
4. Because that container is listening on port 8080, the connection will fail with `connection refused`. 
A connection via `https` would fail for the same reason. 

The solution is to rename the `server1.local` container to something that doesn't correspond to the tld, e.g. `container_server1.local`. 

Side Note: Setting entries in the `hosts` file to `127.0.0.1` will not solve the problem. In that case each container attempting to connect to the server would send things only to itself. This is because from inside a container `127.0.0.1` corresponds to the container itself.
